Matchmaking protocol additions and changes

shutterbug2000 opened the pull request:
- Move utility methods to global so any MatchMaking protocol can use them
- Add EndParticipation to Matchmaking Ext
- Remove clients from Sessions when they leave the server

shutterbug2000 committed:
Matchmaking protocol additions and changes
- Move utility methods to global so any MatchMaking protocol can use them
- Add EndParticipation to Matchmaking Ext
- Remove clients from Sessions when they leave the server

DaniElectra commented on matchmaking/protocol.go:
```diff
@@ -55,5 +57,17 @@ func NewCommonMatchMakingProtocol(server *nex.Server) *CommonMatchMakingProtocol
 	commonMatchMakingProtocol.UpdateSessionHostV1(updateSessionHostV1)
 	commonMatchMakingProtocol.UpdateSessionHost(updateSessionHost)
 
+	if server.PRUDPVersion() == 0 {
+		server.On("Kick", func(packet *nex.PacketV0) {
+			fmt.Println("Leaving")
+			common_globals.RemoveConnectionIDFromAllSessions(packet.Sender().ConnectionID())
+		})
+	} else {
+		server.On("Kick", func(packet *nex.PacketV1) {
+			fmt.Println("Leaving")
+			common_globals.RemoveConnectionIDFromAllSessions(packet.Sender().ConnectionID())
+		})
+	}
```
```suggestion
	server.On("Kick", func(packet nex.PacketInterface) {
		fmt.Println("Leaving")
		common_globals.RemoveConnectionIDFromAllSessions(packet.Sender().ConnectionID())
	})
```

We can use PacketInterface here

DaniElectra commented on globals/matchmaking_globals.go:
```diff
@@ -9,4 +11,60 @@ type CommonMatchmakeSession struct {
+}
+
+func FindClientSessions(clientConnectionID uint32) []uint32 {
+	//TODO: is there any real instance where a client could be in multiple MatchmakeSessions at once? afaik there isn't, but I handle it regardless just in case (if we're sure there isn't, we could remove the handling and speed the code up)
```
The only thing like this that appear on servers are Mario Kart 7 communities, but those use `PersistentGathering`. Then the communities have a list of active MatchmakeSessions, so I'm sure there's no way a client could be in multiple MatchmakeSessions

DaniElectra commented on globals/matchmaking_globals.go:
```diff
@@ -9,4 +11,60 @@ type CommonMatchmakeSession struct {
+	}
+	for _, sessionIndex := range candidateSessionIndexes {
+		sessionToCheck := Sessions[sessionIndex]
+		if len(sessionToCheck.ConnectionIDs) >= int(sessionToCheck.GameMatchmakeSession.MaximumParticipants){
```
```suggestion
		if len(sessionToCheck.ConnectionIDs) >= int(sessionToCheck.GameMatchmakeSession.MaximumParticipants) {
```
missing space

DaniElectra commented on globals/matchmaking_globals.go:
```diff
@@ -9,4 +11,60 @@ type CommonMatchmakeSession struct {
+    return append(s[:index], s[index+1:]...)
+}
+
+func RemoveConnectionIDFromRoom(clientConnectionID uint32, gathering uint32){
```
```suggestion
func RemoveConnectionIDFromRoom(clientConnectionID uint32, gathering uint32) {
```
missing space

jonbarrow commented on globals/matchmaking_globals.go:
```diff
@@ -9,4 +11,60 @@ type CommonMatchmakeSession struct {
+			Sessions[gathering].ConnectionIDs = DeleteIndex(Sessions[gathering].ConnectionIDs, index)
+		}
+	}
+	if len(Sessions[gathering].ConnectionIDs) == 0{
```
```suggestion
	if len(Sessions[gathering].ConnectionIDs) == 0 {
```

missing space

jonbarrow commented on globals/matchmaking_globals.go:
```diff
@@ -9,4 +11,60 @@ type CommonMatchmakeSession struct {
+		}
+	}
+	if len(Sessions[gathering].ConnectionIDs) == 0{
+		Sessions[gathering] = CommonMatchmakeSession{} //TODO: is there a better way to do this? right now, the index is used as the gathering id, and I'm not sure if there's a good way to remove without reordering.
```
If we are using a known key then using a `map` would be faster as the data grows larger. In Go when reading data from either type, a `map` is `O(1)` whereas a slice is `O(n)` based on the size of the slice

jonbarrow commented on globals/matchmaking_globals.go:
```diff
@@ -9,4 +11,60 @@ type CommonMatchmakeSession struct {
+
+func FindClientSessions(clientConnectionID uint32) []uint32 {
+	//TODO: is there any real instance where a client could be in multiple MatchmakeSessions at once? afaik there isn't, but I handle it regardless just in case (if we're sure there isn't, we could remove the handling and speed the code up)
+	gatheringsFoundIn := []uint32{}
```
```suggestion
	gatheringsFoundIn := make([]uint32, 0, len(Sessions))
```

Creating a slice like this is only really useful when we don't know the length the slice will eventually be, and even then `make` can still be used in that case

In this case we know the max length the slice can be, since it can't be any larger than `len(Sessions)`. `make([]uint32, 0, len(Sessions))` will make a uint32 slice with 0 elements but has a max capacity of `len(Sessions)`. Makes it easier for Go to manage this memory

jonbarrow commented on globals/matchmaking_globals.go:
```diff
@@ -9,4 +11,60 @@ type CommonMatchmakeSession struct {
+	return gatheringsFoundIn
+}
+
+func RemoveConnectionIDFromAllSessions(clientConnectionID uint32){
```
```suggestion
func RemoveConnectionIDFromAllSessions(clientConnectionID uint32) {
```

Missing space

jonbarrow commented on globals/matchmaking_globals.go:
```diff
@@ -9,4 +11,60 @@ type CommonMatchmakeSession struct {
+func FindSearchMatchmakeSession(searchMatchmakeSession match_making.MatchmakeSession) int {
+	returnSessionIndex := math.MaxUint32
+	//this portion finds any sessions that match the search session. It does not care about anything beyond that, such as if the match is already full. This is handled below.
+	candidateSessionIndexes := []int{}
```
Same comment as before about using `make` here since we know the max capacity this slice can be

jonbarrow commented on globals/matchmaking_globals.go:
```diff
@@ -9,4 +11,60 @@ type CommonMatchmakeSession struct {
+			candidateSessionIndexes = append(candidateSessionIndexes, index)
+		}
+	}
+	for _, sessionIndex := range candidateSessionIndexes {
+		sessionToCheck := Sessions[sessionIndex]
```
Instead of looping to find indexes, and then also looping over the indexes later, why not just do this logic inside the first loop? Also line break between the `for` loops

shutterbug2000 replied:
I did it this way in case we want to handle more complex matchmaking logic
e.g. instead of just filling the first room we find, search for one with more players, closer VR/whatever skill metric, etc.
so we can find rooms that we can use, and then narrow down the best one

jonbarrow replied:
Gotcha, sounds good then. Shouldn't be a huge performance issue regardless, was just looking for some micro-optimizations and ways to clean things up. But that makes sense

jonbarrow commented on globals/matchmaking_globals.go:
```diff
@@ -9,4 +11,60 @@ type CommonMatchmakeSession struct {
+
+func RemoveConnectionIDFromAllSessions(clientConnectionID uint32){
+	foundSessions := FindClientSessions(clientConnectionID)
+	for gatheringIndex, _ := range foundSessions {
```
If you only need the index, a standard `for` loop is likely more performant since `for range` already reads the value at the index during iteration, which is thrown out here

shutterbug2000 replied:
As in 
`gatheringIndex := 0; gatheringIndex <= len(foundSessions); gatheringIndex++` ? or just removing the _?

jonbarrow replied:
```suggestion
	for i := 0; i < len(foundSessions); i++ {
```

(`i` can be changed that's just my usual default)

jonbarrow commented on globals/matchmaking_globals.go:
```diff
@@ -9,4 +11,60 @@ type CommonMatchmakeSession struct {
+func FindClientSessions(clientConnectionID uint32) []uint32 {
+	//TODO: is there any real instance where a client could be in multiple MatchmakeSessions at once? afaik there isn't, but I handle it regardless just in case (if we're sure there isn't, we could remove the handling and speed the code up)
+	gatheringsFoundIn := []uint32{}
+	for gatheringIndex, _ := range Sessions {
```
If you only need the index, a standard `for` loop is likely more performant since `for range` already reads the value at the index during iteration, which is thrown out here

shutterbug2000 replied:
One last thing and I think this is ready: does this apply now that we're using a map for Sessions? since the keys aren't gonna be in order every time anymore

jonbarrow replied:
No, when using a map type this is the right way to go. `for range` doesn't return an index, it returns a key (it just so happens that with a slice those are the same thing), so this is the right way to go to get these keys

jonbarrow replied:
Though I would say that the name `gatheringIndex` is misleading since it's no longer an index

shutterbug2000 committed:
Code review changes

shutterbug2000 committed:
Change session removal to go with the first session found
There shouldn't be any instance where a client can be in multiple sessions at once

shutterbug2000 committed:
Add a space that I missed
whoops lol
